pkg/clustermanager/eksa_installer.go

```diff
@@ -85,21 +86,20 @@ func (i *EKSAInstaller) Install(ctx context.Context, log logr.Logger, cluster *t
 
 // Upgrade re-installs the eksa components in a cluster if the VersionBundle defined in the
 // new spec has a different eks-a components version. Workload clusters are ignored.
-func (i *EKSAInstaller) Upgrade(ctx context.Context, log logr.Logger, c *types.Cluster, currentSpec, newSpec *cluster.Spec) (*types.ChangeDiff, error) {
+func (i *EKSAInstaller) Upgrade(ctx context.Context, log logr.Logger, c *types.Cluster, currentManagementComponentsVersionsBundle *releasev1alpha1.VersionsBundle, newSpec *cluster.Spec) (*types.ChangeDiff, error) {
 	log.V(1).Info("Checking for EKS-A components upgrade")
 	if !newSpec.Cluster.IsSelfManaged() {
 		log.V(1).Info("Skipping EKS-A components upgrade, not a self-managed cluster")
 		return nil, nil
 	}
-	changeDiff := EksaChangeDiff(currentSpec, newSpec)
+
+	newVersionsBundle := &newSpec.Bundles.Spec.VersionsBundles[0]
```

Comment:
the fact that you have this same line repeated in all upgraders is a good clue that we might be able to find a better abstraction

For example, what if your interface looks like this?

```go
 Upgrade(ctx context.Context, log logr.Logger, c *types.Cluster, current, new *releasev1alpha1.VersionsBundle) (*types.ChangeDiff, error) {
 ```
 
 (you might need to include the spec or the cluster and that's fine but cleaner if not)
 
This is more explicit: it's given the caller control over which version bundle should be the new one. The caller can orchestrate all the upgraders to use the same version bundle instead of expecting all of them to behave in sync.

I would apply this to all upgraders.